client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx

```diff
@@ -18,140 +16,157 @@ import {
 } from '../../common/styleConstants';
 import { useSearchQueries } from './useSearchQueries';
 import { UserOptions } from './Search';
-
-const MIN_CHARS = 3;
+import { JsonFormsDispatch } from '@jsonforms/react';
 
 export const SearchWithUserSource = (
   props: ControlProps & { options: UserOptions }
 ) => {
-  const { data, path, handleChange, label, visible, options } = props;
+  const {
+    data,
+    path,
+    handleChange,
+    label,
+    visible,
+    options,
+    schema,
+    renderers,
+  } = props;
   const t = useTranslation('programs');
-  const [searchText, setSearchText] = useState('');
-  const [editMode, setEditMode] = useState(!data);
-  const [noResultsText, setNoResultsText] = useState(
-    t('control.search.searching-label')
-  );
+
+  const isPatientSelected = !!data?.id;
 
   const {
     runQuery,
     getOptionLabel,
-    getDisplayElement,
     saveFields,
-    placeholderText,
-    loading,
     error: queryError,
     results,
+    resetResults,
   } = useSearchQueries(options ?? {});
 
-  const debouncedOnChange = useDebounceCallback(
-    value => {
-      if (value.length >= MIN_CHARS) runQuery(value);
-      else {
-        // Clear the results if user input falls *below* `minChars`
-        if (results.length) runQuery('');
-      }
-      setNoResultsText(t('control.search.no-results-label'));
-    },
-    [searchText],
-    500
-  );
-
-  const handleDataUpdate = (selectedResult: Record<string, unknown> | null) => {
-    if (selectedResult === null) return;
-    if (!saveFields) handleChange(path, selectedResult);
-    else {
-      const newObj: Record<string, unknown> = {};
-      saveFields?.forEach(
-        field => (newObj[field] = selectedResult[field] ?? null)
-      );
-      handleChange(path, newObj);
+  useEffect(() => {
+    if (!data) {
+      handleChange(path, {});
+      return;
     }
-    setEditMode(false);
-  };
 
-  const getNoOptionsText = () => {
-    switch (true) {
-      case loading:
-        return t('control.search.searching-label');
-      case searchText.length < MIN_CHARS:
-        return t('control.search.below-min-chars', { minChars: MIN_CHARS });
-      default:
-        return noResultsText;
+    // Only run the database query if a specific patient hasn't yet been
+    // selected
+    if (isPatientSelected) return;
+
+    const searchFilter: FilterBy = {};
+    options.searchFields.forEach(field => {
+      const match =
+        field in queryMatchTypes
+          ? queryMatchTypes?.[field as keyof typeof queryMatchTypes]
+          : 'like';
+      if (data[field]) searchFilter[field] = { [match]: data[field] };
+    });
+    if (Object.keys(searchFilter).length > 0) runQuery(searchFilter);
```

Comment:
I think it's better/easier to use a TanStack `useQuery` instead of a direct API call. In useQuery you have the option to enable the query, e.g. you can enable it when `isPatientSelected` is true. Advantage is that you don't have to do use a `useEffect` and you just pass in the current filter parameters and get the updated results when they are ready. Furthermore, it does some caching.

For example, similar to:
`packages/system/src/Patient/api/hooks/document/usePatient.ts`

Reply:
I don't really get this, I'm finding the labyrinth of nested hooks and apis too convoluted to navigate clearly, and I can't seem to create a simple hook that gives me the same search as I have currently. I'll have to get you to show me how to make sense of it when you're in the office next.

Reply:
yes sure can show you tomorrow. Its basically just wrapping the runQuery call in a useQuery call...

Btw. @mark-prins and I had some ideas to simplify the hook/api setup, maybe next R&D day...

Reply:
> Btw. @mark-prins and I had some ideas to simplify the hook/api setup, maybe next R&D day...

Ooh nice! I look forward to hearing more about that.

Reply:
Righto, I've done that now, and I've simplified the internal hook to only handle the one query type. It's much simpler now: 3696c298a

Reply:
> Ooh nice! I look forward to hearing more about that.

One idea I had was generating the react-query part as well:
https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-react-query

If this wouldn't work I thought about just simplifying the dir structure...